internal/core/software_renderer.rs

```diff
@@ -1146,23 +1242,124 @@ struct RenderState {
     clip: LogicalRect,
 }
 
+fn round(l: Length<f32, PhysicalPx>) -> Length<i16, PhysicalPx> {
+    Length::new(l.get().round()).cast()
+}
+
 impl<'a, T: ProcessScene> crate::item_rendering::ItemRenderer for SceneBuilder<'a, T> {
     fn draw_rectangle(&mut self, rect: Pin<&crate::items::Rectangle>, _: &ItemRc) {
         let geom = LogicalRect::new(LogicalPoint::default(), rect.geometry().size_length());
         if self.should_draw(&geom) {
-            let geom = match geom.intersection(&self.current_state.clip) {
+            let clipped = match geom.intersection(&self.current_state.clip) {
                 Some(geom) => geom,
                 None => return,
             };
 
+            let background = rect.background();
+            if let Brush::LinearGradient(g) = background {
+                let geom2 = geom.cast() * self.scale_factor;
+                let clipped2 = clipped.cast() * self.scale_factor;
+                let act_rect = (clipped.translate(self.current_state.offset.to_vector()).cast()
+                    * self.scale_factor)
+                    .round()
+                    .cast();
+
+                let angle = g.angle();
+
+                let tan = angle.to_radians().tan().abs();
+                let start = if !tan.is_finite() {
+                    255.
+                } else if false && geom.width() > geom.height() {
+                    let h = tan * geom.height();
+                    255. * h / (h + geom.width())
+                } else {
+                    let h = tan * geom.width();
+                    255. * h / (h + geom.height())
+                } as u8;
+                let mut angle = angle as i32 % 360;
+                if angle < 0 {
+                    angle += 360;
+                }
+                let mut stops = g.stops().copied().peekable();
+                let mut idx = 0;
+                let stop_count = g.stops().count();
+                while let (Some(mut s1), Some(mut s2)) = (stops.next(), stops.peek().copied()) {
```

Comment:
This creates one item scene item for each stops pair instead of one command with the stops.
I thought it would be better to do that in order to avoid allocating a stop array in the commands. for the common case of only 2 stops.

But maybe that's a wrong choice and we can probably find a way to efficiently copy the stops over (since it is a SharedVector)
Also, i currently have some artifacts at the transition that aren't great.

Reply:
Makes sense to me to have one per command instead of stops pair and take a cheap clone of the `LinearGradientBrush`.

Reply:
I'm going to keep it like this for now and we can optimize that later.

Reply:
Ok